### ISSUE TITLE
New Module System Discussion - Please contribute.

### ISSUE THREAD

**Issue by MichMich** [MichMich/MagicMirror#80]
Maybe it's a good idea to discuss the new [Module System](https://github.com/MichMich/MagicMirror/wiki/Module-System-Proposal) before we writeout the API. 

@paviro, I noticed you've contributed. Thanks for this. I really appreciate your help! Just a few questions though: 

The functions you proposed, are those method of a module class? I can't really wrap my head around how this would work, could you elaborate?

Also, I think the added regions are a bit unnecessary. In most cases you can just add content in one of the corners (the content will stack) or align right or left in the upper third, center or lower third. I think we should try to keep the layout simple.

Aslo, what do you guys think about using a system like vue.js? I'm not a big fan of jQuery, and prefer a mvvm model. But this might make it to difficult for most users to contribute. 

**EDIT 2016/3/23: First steps in the development are made. See: https://github.com/MichMich/MagicMirror/issues/80#issuecomment-200411278**

**Comment by paviro** [comment]
To be honest with you I haven't thought all that much about how to implement it. I just added what I would need to port my modules to the new system, which is:

- Helper scripts written in Python and Node.js running as background daemons
- A way to switch between multiple views (one for general users and one for people who are logged in via facial recognition). I would therefore love the possibility to be able to define more than one module layout in config.js and also to switch them via a module. Might also be useful if you want to have different views for lets say work related and private stuff. A module could again provide an interface to switch between them for example via a GPIO button.

Since my edits in the wiki are not that useful in their current state, I will remove them, they are hear now anyway.

**Comment by bitte-ein-bit** [comment]
This is going to be some random thoughts.

I think if you switch to vue.js most user will not be able to contribute. I'm not sure about stability. I once used a rather new php framework (symfony 1.x) and later regretted it because their api change a LOT when going to 2.0. IMHO jQuery seems to have stabilized. I can't tell for vue.js or angular.js. Never used either. On the other hand not using spaghetti code is nice.

Personally I think it would be a good idea if a module could listen and trigger events. Some events could be defined (e.g. 'User changed', 'User logged out', 'Calendar event starting', 'GPIO x HIGH', ...) and some central instance could route them. This core should be very minimal and all that is needed to run the mirror. I think that everything else should be a module.

I agree with @paviro that some modules should be implemented without any use by themself like `socket.io`. This makes it necessary that modules can specify depends.

Regarding the helper scripts I think it would be nice to have e.g. a Makefile which will download and install the dependencies and setup an init.d or systemd script. Writing an daemon is rather advanced so IMHO the developer of that plugin should be able to find enough information to provide a Makefile. Other options would be some sort of Chef cookbook, Puppet Manifest, Bash Script, ...  which could also setup the system itself. A Makefile would probably be the easiest.

I think the regions will work fine. But we should find a way, that a user can override all region positions without interfering with the main code. Not all mirrors are the same. So they might need some way to adjust margins, etc. For some modules it would be nice to render depending on the screen orientation but that is just eye candy.

**Comment by paviro** [comment]
I think I will write a Python CLI which could be used to install modules and also list all available modules with a description. I could implement the systemd or init.d scipt setup there :)

## CLI
Something like:
- magic-mirror list (show all plugins with short discription)
- magic-mirror info module (show long description, dependencies etc. for module)
- magic-mirror install module (installs module and dependencies)
- magic-mirror remove (removes module and dependencies, if installed via magic-mirror install)
- magic-mirror enable/disable module (?)
- magic-mirror update (Update module list and versions)
- magic-mirror upgrade (install updates)

**Comment by MichMich** [comment]
Hi @paviro, before we start building command line tools, It might be a good idea to first finish of the basic concept behind the modules. How they work, and how dependencies are loaded. - Also, since most of it is javascript based, we might consider using existing CLI tools, like npm. This way, we can put our effort in the enhancement of the mirror UI.

**Comment by MichMich** [comment]
@bitte-ein-bit I agree with the idea of event triggering. Besides a module instance, there should be indeed a MagicMirror shared instance (singleton concept) which can receive messages of the module instance (in other words, a delegate for the module). 

The MagicMirror share instance (the main instance) could have methods to load and unload modules. A broadcasting system for notifications between modules (so no direct module to module communication). A method to reload the HTML/dom for a specific module ... etc.

Regarding the overriding for region positions: this can all be done by css. The region is just a simple way to load a module in a specific location, but the module is free to adjust it's design using CSS. (As long as it doesn't interfere with other visuals.)

CSS based, it might be a good idea to define some standard elements modules can be used. If all modules use the same css elements, we could also implement a tempting system in later stage. 

Regarding dependencies and installing libs. I think this should be taken case of if we have the base of the module system running. IMHO we could use existing systems like NPM.

Most importantly: thanks for your effort and contributions. I love the fact we are working on the future of the Magic Mirror.

**Comment by MichMich** [comment]
As a small side note: If we start using NPM, we could also consider building a small node web server to host the magic mirror. This way, users don't need to install Apache & PHP.

**Comment by gefangenimnetz** [comment]
I would love a JS only implementation. That would allow for easy porting to to Android or IOS using build tools like cordova (phonegap), etc. Basically that’s what I am currently doing.

**Comment by paviro** [comment]
@MichMich Your are right :)

**Comment by MichMich** [comment]
Not sure yet, but I might have some spare time end of this month to work on this version.

**Comment by sclausen** [comment]
@gefangenimnetz I'm currently working on this. It's based on angular2 and I'm still not sure, if I'll add a backend. Meteor (still with an angular2 frontend) is one idea and works really well, especially with realtime server side functionality like motion detection.

**Comment by paviro** [comment]
@MichMich  @bitte-ein-bit  any news on your side? Any way I can help you with the module system with my limited JS knowledge?

**Comment by sclausen** [comment]
@gefangenimnetz @MichMich I started a javascript-only clone [sclausen/MagicalMirror](https://github.com/sclausen/MagicalMirror)

**Comment by bitte-ein-bit** [comment]
Unfortunately I'm packed with other stuff to do. Don't have the time to
lock into a new backend right now. Sorry.

Am Do, 10. Mär 2016, um 14:46, schrieb Sebastian Clausen:

**Comment by sclausen** [comment]
@bitte-ein-bit currently it has no backend and works frontend only with calendar, news, weather and time.

**Comment by MichMich** [comment]
I'll be probable be working on a new version coming weeks. I'll keep you guys posted.

**Comment by paviro** [comment]
Nice looking forward to it!

**Comment by MichMich** [comment]
Hi Guys,

Today, I FINALLY started working on the module system. Although I only have the very simple foundation up and running, I think I'm really on the right track.

A small overview so far:

**config.js**
Used to define all the modules that need to be loaded, including the configuration for the specific module.
Adding a module to this config will ensure the module will be loaded, including all nessecery .js and .css files. So other then downloading a module to the module folder and adding it to the config, no further actions are required to add it to the mirror.

```javascript
	var config = {
		modules: [
			{
				module: 'clock',
				position: 'middle_center',
				config: {
					foo: 'Bar'
				}
			},
			{
				module: 'compliments',
				position: 'middle_center',
				config: {
					compliment: 'You look hot!'
				}
			},
			{
				module: 'compliments',
				position: 'top_right',
			}
		]
	};
```
**modules/compliments/compliments.js**
This is a very simple example of a module. Of course, this doesn't include any logic yes, but shows how simple a module is. 

```javascript
	Module.create({

		// Define the default config. This is optional.

		defaults: {
			compliment: "This is the default compliment"
		},


		// Define a start method. For example if you want to schedule timers,
		// or want to make connections to socket servers. Again: this is optional.

		start: function() {
			Log.info('Starting module: ' + this.name);
		},


		// Return an array of scripts that need to be loaded.
		// All scripts are loaded before the module is started.
		// Loading might be skipped if the file is already loaded.
		// The system contains a vendor folder for default libs like jQuery.
		// See comments below for more info.
		// This method is optional.

		getScripts: function() {
			return [
				'dummy.js', // Loaded (if available) form vendor folder, otherwise from module folder. Loaded only once.
				'http://server.com/js/file.js', // External file. Always loaded.
				this.file('foo.js'), // File from module folder. Multiple foo.js files might be loaded.
			];
		},


		// Same system as getScripts().
		// This method is optional.

		getStyles: function() {
			return [
				'compliments.css'
			];
		},


		// Returns the dom to be inserted in the mirror.
		// This method is optional.

		getDom: function() {
			var compliment = document.createTextNode(this.config.compliment);       
			var div = document.createElement("div");
			div.appendChild(compliment);

			return div; 
		}
		
	});
```

And thats all there is to write a module! Of course, I will continue to work on this. If I have any news I'll let you know. Hope to share a basic version next week. 

All code so far is pure javascript. But of course, modules are free to use additional libs like jQuery.

Open for feedback.

**Comment by CFenner** [comment]
Looks great. I like the js only approach.

**Comment by MichMich** [comment]
And, as an example, a super simple module to show something on the mirror:

**modules/helloworld/helloworld.js:**

```javascript
Module.create({
	defaults: {
		text: "Hello World!"
	},
	getDom: function() {
		return document.createTextNode(this.config.text);  
	}
});
```

This allows you to display a text on the mirror via the config: 

```javascript
var config = {
	modules: [
		{
			module: 'helloworld',
			position: 'middle_center',
			config: {
				text: 'This works perfect!'
			}
		},
	]
};
```

**Comment by paviro** [comment]
Looks great! :) How are we going about socket connections? Should each module create their own (also if it uses helper scripts) or should we create a connection module which serves information from one socket to all the modules which need it and helper scripts have to talk to our socket?

**Comment by MichMich** [comment]
I've implemented a notification system to communicate between modules. This will also be used for the incoming socket messages, and probably also for the outgoing messages (although I don't see a use case for outgoing messages yet ...).

**Comment by paviro** [comment]
There is a use case. I would like to implement voice control and also use that for example to set my alarm clock which would need outgoing connections.

**Comment by MichMich** [comment]
But in that case you might want to communicate with a socket other then the mirror's backend. For now, the backend won't include a module system. (Will definitely be on the todo list.)

**Comment by MichMich** [comment]
I pushed a very simple first version of the v2-beta to a separate branch. Unfortunately I did not have the time yet to write a readme, but most files have sufficient comments. I probably won't be able to work on this version coming few days, but will definitely continue the project next week.

Feel free to give it a spin!

https://github.com/MichMich/MagicMirror/tree/v2-beta

**Comment by Miyasashi** [comment]
Can someone explain what the module system does I dont really get it.

**Comment by MichMich** [comment]
It will (eventually) allow everyone to built modules for the magic mirror without the need of modifying the core of the system. This way everyone can provide modules that can be installer on the mirror.

**Comment by Miyasashi** [comment]
Thanks for the fast response, funny coincidence i had problems with implenting my own stuff and I think this will help a lot.

**Comment by paviro** [comment]
@MichMich Do you think we could move the whole project to [Electron](http://electron.atom.io/) once the module API is finished? Only the calendar part would have to be ported to JS right?

**Comment by CFenner** [comment]
It's not so easy to get a more complex module to run. This variable structure differs and Q seems not to work.

**Comment by MichMich** [comment]
@Paviro: Maybe in the future. It's not the current focus. 
@CFenner: I don't really understand what You mean. Could you elaborate?

**Comment by CFenner** [comment]
As the module was accessible via the global variable it was a bit easier to chain up the methods with Q promises. Now I need to bind the correct module object to each handler method to be able to call the next method. But I think I'm making progress..

**Comment by MichMich** [comment]
Hi @CFenner, I don't have any experience with 'Q promises'. Could you give an example what was possible in the old version, which isn't in the new version. Also note: the new version is in a very early state. So a lot of things might break in future versions, so please don't spent to much time on building stuff for the new version yet. :)

**Comment by CFenner** [comment]
Before I had this: 
```
return Q.fcall(
	netatmo.request.token, netatmo.render.error
).then(
	netatmo.request.data, netatmo.render.error
).then(
	netatmo.render.all
).then(
	netatmo.update.wait
);
```

Now it looks like this:
```
var that = this;
return Q.fcall(
	this.request_token.bind(that), this.render_error.bind(that)
).then(
	this.request_data.bind(that), this.render_error.bind(that)
).then(
	this.render_all.bind(that)
).then(
	this.update_wait.bind(that)
);
```

But I think it is needed because Q is using the methods as a callback from the ajax request, but after all it's working now. 

<img width="721" alt="bildschirmfoto 2016-03-25 um 18 04 37" src="https://cloud.githubusercontent.com/assets/9592452/14049247/45058c36-f2b4-11e5-98bb-d0804ea6b55a.png">

**Comment by MichMich** [comment]
@paviro the node_helper todo's are all gone. :)

What's more important, is that I integrated a socket system for the node helpers. Using it is pretty simple:

**In client side module:**
To send a notification:
```javascript
var notification = "JUST_A_STRING"
var payload = {anything: 'you want'}
this.sendSocketNotification(notification, payload);
```

To receive messages, make sure your client side module has the following method:
```javascript
socketNotificationReceived: function(notification, payload) {
	console.log(notification);
	console.log(payload);
},
```

**In node_helper.js:**
Create a socket object: 
```javascript
var MMSocket = require('../../js/socketclient.js');
var socket = new MMSocket('nameofthemodule');
```
To send a notification:
```javascript
var notification = "JUST_A_STRING"
var payload = {anything: 'you want'}
socket.sendNotification(notification, payload);
```
To receive messages, register a callback:
```javascript
socket.setNotificationCallback(function(notification, payload) {
	console.log(notification);
	console.log(payload);
});
```

**Note 1:** The connection will be established as soon as the client side sends it's first message. 
**Note 2:** All modules of the same type receive the same notifications. 
**Note 3:** For the client side modules, we're using subclassing to keep thing more simple. (No need to create a socket object for instance ...), I might work on a node_helper.js superclass as well in order to streamline the experience.

**Comment by paviro** [comment]
Nice! I will later change my module to use it :)
The weather module looks really gorgeous!

**Comment by MichMich** [comment]
Ok, slight change. I changed the node_helper setup to use a superclass. This way it works the same as the clients side modules, So to recap: 

To send a notification:
```javascript
var notification = "JUST_A_STRING"
var payload = {anything: 'you want'}
this.sendSocketNotification(notification, payload);
```

To receive messages, make sure your client side module or your node_helper has the following method:
```javascript
socketNotificationReceived: function(notification, payload) {
	console.log(notification);
	console.log(payload);
},
```

**Comment by MichMich** [comment]
About that namespace: it doesn't create a lot of overhead, but never the less it will be fixed in the future. Currently all node helpers connect to the socket with their own connection. In the future they just use the overal socket connection. This will also fix the issue you mentioned.

This should not affect any of your code.

**Comment by paviro** [comment]
Alright :+1:

**Comment by paviro** [comment]
Do you think it would make sense to add some kind of notification UI module, so that modules can show alerts? I use [sweetalert](https://github.com/t4t5/sweetalert) for my callmonitor but maybe it would make sense to integrate something like that into the core system? I am thinking of slide-in notifications (OS X style) but also UI overlays like I use in the monitor.

**Comment by MichMich** [comment]
Alerts: this is exactly what I was thinking. This would be a nice feature for the future.
Swap module: yes this is on the to do list.

**Comment by CFenner** [comment]
An alerts system would be a nice idea so each module can display configuration or communication issues in a similar way. 

Maybe build a module for it and use the socket system.

**Comment by MichMich** [comment]
There already is a inter-module notification system. So we just need to built this module.

To send a notification to all modules (client side, so no socket connection):
````javascript
sendNotification: function(notification, payload);
````

To receive notifications:
````javascript
notificationReceived: function(notification, payload, sender) {
		if (sender) {
			Log.log(this.name + ' received a module notification: ' + notification + ' from sender: ' + sender.name);
		} else {
			Log.log(this.name + ' received a system notification: ' + notification);
		}
	},
````

**Comment by MichMich** [comment]
@paviro: That namespace issue is now solved. https://github.com/MichMich/MagicMirror/commit/b243d25399e9538a0dc563e3fd7dd58c864937c1

**Comment by paviro** [comment]
Nice! Glad module swapping is on the todo list!

**Comment by MichMich** [comment]
What would be the best approach in your oppinion?

My idea is to allow the configuration of additional module classes in the config.

````javascript
var config = {
	modules: [
		{
			module: 'calendar',
			position: 'top_left',
			classes: 'michmich'
		},
		{
			module: 'calendar',
			position: 'top_left',
			classes: 'paviro'
		}
	]
}
````
This way you could do stuff like: 

````javascript
// using the default classes
MM.hideModulesWithClass('calendar',  speed, callbackFunction);

//using the custom classes
MM.hideModulesWithClass('michmich',  speed, callbackFunction);
MM.showModulesWithClass('paviro',  speed, callbackFunction);
````

Any ideas?

**Comment by paviro** [comment]
I like that way :)

**Comment by MichMich** [comment]
Cool, will try to implement it asap.

**Comment by paviro** [comment]
Nice, I will try to finish my facial recognition module later tonight.

**Comment by MichMich** [comment]
Great idea. Will keep it in mind.

Calendar module is almost finished. Will push it tomorrow.

**Comment by paviro** [comment]
Nice! Is there currently a standard way of reading a modules config from the node helper script? Other than by require it?

**Comment by MichMich** [comment]
I thought about it, but since a module can be included more then once, it has no use to push the config of the client to the node_helper. None the less, you should be label to just use the `config` variable (i think it's available in for the helper as well since it's a global ...), best practice would be to send the configuration from a module using the sockets when you perform a socket request.

**Comment by paviro** [comment]
Yeah but my face reg module should not be loaded more than once anyway would not make any sense. Does not seem like I can access it via `this.config` just returns `undefined`. Would be great if it would be possible, since the module will use a python script and I want to allow the user to set the settings via the `config.js`.

**Comment by MichMich** [comment]
You could try just `config` (without this.).

I think the best approach would be:

modulename.js
````javascript
start: function() {
    sendNotification: function('CONFIG', this.config);
}
````

node_helper.js
````javascript
socketNotificationReceived: function(notification, payload) {
     if (notification === 'CONFIG') {
         this.config = payload;
         //start setup of the python object.
    }
}
````

**Comment by paviro** [comment]
I can't figure out what I am doing wrong...
I used `sendNotification: function('CONFIG', this.config);` but than I got an error` Uncaught SyntaxError: Unexpected token (`. If I do `this.sendNotification('CONFIG', this.config);` nothing ever gets received in the node helper..

**Comment by MichMich** [comment]
Could you post your code somewhere?

**Comment by MichMich** [comment]
It looks ok. Can't really check it since I am on my phone. Did you pull the latest v2 version?

**Comment by MichMich** [comment]
Also, why use the timer? You could start python when the config arrives. (IF it arrives. ;))

**Comment by paviro** [comment]
I used the timer because I was scared that it could be started twice if someone else sends something as CONFIG...
I updated now, have you changed how the helpers are called? Because I now get `No helper found for module: MMM-FRITZ-Box-Callmonitor.`.

**Comment by paviro** [comment]
Works now, updating did the trick, thanks!

**Comment by MichMich** [comment]
So it's solved? 


Also, do something like:

If(!pythonStarted) {
   pythonStarted=true;
   pythonStart();
}

**Comment by paviro** [comment]
That makes sense! Is it correct that I can't access `this.name` within `defaults`? I am trying to set the default path to the training file and therefore add the modules name to the path but it seems to be `undefined`.

**Comment by MichMich** [comment]
Correct. It's not set yet when instanciating the module. Going to bed now. Will continue further dev tomorrow.

**Comment by paviro** [comment]
Alright then I'll hardcode it for now. Sleep well :)

**Comment by MichMich** [comment]
I made a breaking change to the module loader. Modules should now be registered using the following method:

Old way:
````javascript
Module.create({
 ...
});
````

New way:
````javascript
Module.register('helloworld',{
 ...
});
````

**Comment by paviro** [comment]
Updated my modules. What is the advantage internally of loading them this way?

**Comment by MichMich** [comment]
Easier debugging. And no need for tricky `eval()` method.

I'm currently experiencing an issue though. One module won't be started. Currently investigating it.

**Comment by MichMich** [comment]
Issue is fixed.

**Comment by paviro** [comment]
Any ETA on the module swapping?

**Comment by MichMich** [comment]
Working on it as we speak ... ;)

**Comment by MichMich** [comment]
Ok, so, the module swapping should be working now ... It needs some documentation, so here you go.

To make a selection of modules, you have a lot of options:
````javascript
// All modules:
var module = MM.getModules();

// All modules with specific classes:
var module = MM.getModules().withClass('classname');
var module = MM.getModules().withClass('classname1 classname2');
var module = MM.getModules().withClass(['classname1','classname2']);

// All modules except with specific classes:
var module = MM.getModules().exceptWithClass('classname');
var module = MM.getModules().exceptWithClass('classname1 classname2');
var module = MM.getModules().exceptWithClass(['classname1','classname2']);

// All modules except a specific module:
var module = MM.getModules().exceptModule(aModule);

// You can combine the methods above:
var module = MM.getModules().withClass('classname1').exceptwithClass('classname2').exceptModule(aModule);
````

If you want to perform an action with those methods, you can use the `enumerate` function:
````javascript
MM.getModules().enumerate(function(module) {
    Log.log(module.name);
});
````

Hiding/showing a module is simple:
````javascript
    //Hide with last used speed.
    module.hide();

    //Hide with 1 second animation.
    module.hide(1000);

    //Hide with 1 second animation and callback.
    module.hide(1000, function() {
        Log.log('Module is hidden.');
    });

    //Show with last used speed.
    module.show();

    //Show with 1 second animation.
    module.show(1000);

    //Show with 1 second animation and callback.
    module.show(1000, function() {
        Log.log('Module is shown.');
    });
````

And of course, you can combine all of the above:
````javascript
MM.getModules().withClass('classname1').exceptwithClass('classname2').exceptModule(aModule).enumerate(function(module) {
    module.hide(1000, function() {
        Log.log('Module is hidden.');
        module.show(1000, function() {
            Log.log('Module is shown.');
        });
    });
});
````

Last but not least, if you want to add a class to a specific module in the config, this can be done using the `classes` property:
````javascript
{
	module: 'modulename',
	position: 'lower_third',
	classes: 'classname1 classname2 classname3'
}
````
**Note 1:** The system does not invoke a suspend method on the module yet.
**Note 2:** The method `MM.getModules()` returns an empty array if not all modules are started yet. If you want to make sure all modules are started, you can listen for a system notification.
````javascript
notificationReceived: function(notification, payload, sender) {
	if (notification === 'ALL_MODULES_STARTED') {
		// ... do something with here MM.getModules()
	}
}
````

Let me know how this system works for you.

**Comment by paviro** [comment]
Great nice! I just tried to write a Wunderlist module but am a bit stuck there so I will implement this one now :) I hate asynchronous code (at least right now)...

**Comment by MichMich** [comment]
Lol. I understand what you mean. :)

**Comment by paviro** [comment]
Mhh seems like I am doing something wrong... 
`Uncaught TypeError: MM.getModules(...).withClass(...).exceptwithClass is not a function`
Just copy and pasted what you posted here.

**Comment by MichMich** [comment]
exceptWithClass (capital W)

**Comment by paviro** [comment]
Seems to work! Nice! Only thing, I am using this code to hide any module that does not have the class "default", but it does not do anything.
`if (notification === 'ALL_MODULES_STARTED') {
			MM.getModules().exceptWithClass('default').enumerate(function(module) {
				module.hide(0, function() {
					Log.log('Module is hidden.');
				});
			});
		}`

**Comment by MichMich** [comment]
Above code looks ok and iCloud agenda works here. Ill look into tomorrow if you can't solve it.

**Comment by MichMich** [comment]
I suggest you do some logging with your above code to see which modules it selects.

**Comment by SVendittelli** [comment]
The calendar only seems to load the first of recurring events and only if that event is in the future, see example below

![image](https://cloud.githubusercontent.com/assets/17616742/14188339/9eaac2d6-f77f-11e5-9d60-9860843a48ba.png)
![screenshot 2016-03-31 20 30 33](https://cloud.githubusercontent.com/assets/17616742/14188357/b445bccc-f77f-11e5-8ef8-860876943c3b.png)

**Comment by paviro** [comment]
I do, just added `console.log(module);` and every module is logged just not faded out.. I'll check my `module.hide` again.

**Comment by MichMich** [comment]
@SVendittelli thanks! Could you make a separate issue for this? I'll look into it tomorrow.

**Comment by SVendittelli** [comment]
Sure thing!

**Comment by paviro** [comment]
Could not find the problem yet but I already noticed one thing. Hidden modules take up space on the UI, so if you would for example have three sets of modules for three people that would not work...

**Comment by MichMich** [comment]
Will look into it tomorrow. Could you create a simple module with the non working code so I can test?

**Comment by paviro** [comment]
Sure!

**Comment by MichMich** [comment]
> Hidden modules take up space on the UI, so if you would for example have three sets of modules for three people that would not work...

This should be fixed now.

**Comment by paviro** [comment]
Nice, works thanks! Weirdly the fade in animation seems not to work anymore, modules just appear and then jump to the new position.

**Comment by MichMich** [comment]
Is working here ... REALLY going to bed now. ;)

**Comment by paviro** [comment]
Any idea why this module overlaps with the calendar?
![Screenshot](https://cloud.githubusercontent.com/assets/992826/14194663/4617c592-f7b1-11e5-8de2-c165f94fe8ac.png)

**Comment by MichMich** [comment]
Are they in the same position? If so please open a separate issue with example code.

**Comment by MichMich** [comment]
Found the issue. It's fixed. As well as the fading that didn't work at all times.

**Comment by paviro** [comment]
Nice!

**Comment by MichMich** [comment]
I made a small start with the module documentation:
https://github.com/MichMich/MagicMirror/tree/v2-beta/modules/clock
https://github.com/MichMich/MagicMirror/tree/v2-beta/modules/calendar

I will continue to add documentation of the other models on a later stage. If anyone feels the urge to document the other modules, feel free! ;)

**Comment by MichMich** [comment]
All module README.md's are ready. :)

**Comment by CFenner** [comment]
Is the module system in a 'nearly complete' state by now?

**Comment by MichMich** [comment]
Can't promise anything ... ;)

I'm working on one major change now. But that won't be a breaking change. After that, the next step will be the full documentation.

**Comment by paviro** [comment]
Do you have time to add the UI notifications or should I give it try this evening? :)

**Comment by MichMich** [comment]
No, sorry. No time for it. You can try, but I can't promise I will include/use it. Make sure you just make a separate module for it. (No core changes should be necessary);

**Comment by MichMich** [comment]
I made a few huge changes on the loader. You should now be able to put modules in folder.
In the config you should be able to use `module: 'foldername/modulename'`. The default modules don't need the `default/` prefix.

This way you can make packages of multiple modules.

**Comment by CFenner** [comment]
Do I need to wrap a module in an extra folder or does the old way still work?

**Comment by paviro** [comment]
Still works :)

**Comment by CFenner** [comment]
thanks, just found the error :)

**Comment by konzeptplus-gmbh** [comment]
Hi, 

@MichMich Thanks for your great Codes. 
Do you have an idea how I can add a calender with an auto-login? 
As I'm not a Java-scripter I do not know how I should add the html code in the Java. 

`<HTML>
<HEAD>
<TITLE>Cal</TITLE>
<script>
    function loginForm() {
        document.myform.action = "https://ox.hostkonzept.ch/ox6/#m=calendar&f=26";
        document.myform.submit();
    }
</script>
</HEAD>
<BODY onLoad="loginForm()">
    <FORM NAME="myform" METHOD="POST">
        <INPUT TYPE="hidden" NAME="email" VALUE="myemail"> 
        <INPUT TYPE="hidden" NAME="password" VALUE="mypassw">
    </FORM>
</BODY>
</HTML>`

Thanks for the help. :-)

**Comment by CFenner** [comment]
@konzeptplus-gmbh, this would be better discussed in a separate issue.

**Comment by konzeptplus-gmbh** [comment]
Yes you are right. Sorry. :-)

**Comment by 3vidar** [comment]
Hi guys

I love this project and decided to build my own mirror. But I'm not a coder;o( because of that i just would ask you pros if it is possible to implement in de middle of the screen a smart calendar? just like this shown in the picture (just the calendar connected to google calendar)?
![fa6yc5bilche9z8 large](https://cloud.githubusercontent.com/assets/17111989/14345375/ca892776-fcac-11e5-8c31-503714d0314e.jpg)

![sglpazb6nupgo31sjxjt](https://cloud.githubusercontent.com/assets/17111989/14345232/20a44ede-fcac-11e5-9da0-ca034484ba45.jpg)

Source of the pictures: http://www.instructables.com/id/Raspberry-Pi-Wall-Mounted-Calender-and-Notificatio/

**Comment by MichMich** [comment]
The new system has a calendar module, but not in the way the above calendar looks like. It could easily be built, but as you said, you're not a coder ... ;)

**Comment by 3vidar** [comment]
Because of that I ask you coder if you could build something like this for me? ;o) please

**Comment by MichMich** [comment]
FYI: I've just added the following feature:

> In some cases, you want to start the application without an actual app window. In this case, exectute the following command from the MagicMirror folder: `node server only`. This will start the server, after which you can open the application in your browser of choice.

https://github.com/MichMich/MagicMirror/commit/a34a5c4db087ae7a17261e2053d0ba2bdc8fc08f

**Comment by MichMich** [comment]
I started working on the developer documentation:
https://github.com/MichMich/MagicMirror/tree/v2-beta/modules/README.md

**Comment by MichMich** [comment]
I think most of the documentation is done. Please review.

**Comment by paviro** [comment]
Looks good, nice work! :) Spotted some typos but I have to learn right now.. Will fix them later if nobody else does until then.

**Comment by MichMich** [comment]
Guys, I've just announced this version on my blog: http://michaelteeuw.nl/post/142629029062/magicmirror²-a-huge-announcement

Including a special thank you to you guys! :)

**Comment by paviro** [comment]
Nice one! Happy I could help and thanks again for this now even more awesome project!

**Comment by MichMich** [comment]
Since the new version is up and running, I close this issue. Please open separate issues for new feature requests and bugs.
